initiate CPUTensor.

ceciliapeng2011 opened the pull request:
### Details:
 - *Write ITensor implementation over cpu Memory (presumably it is better to write it over IMemory) and the necessary unit tests (can be stollen from the Core team), thanks to IMemory with a mock memory class. Merge into dev branch. *
 - *...*

### Tickets:
 - *ticket-id*

maxnick commented on src/plugins/intel_cpu/src/cpu_tensor.h:
```diff
@@ -0,0 +1,50 @@
+
+    void* data(const element::Type& type = {}) const override;
+
+    template <typename T, typename datatype = typename std::decay<T>::type>
+    T* data() const {
+        return static_cast<T*>(data(element::from<datatype>()));
+    }
```
Do we really need to redefine this method? It looks like the base one should be enough.

ceciliapeng2011 replied:
I have to, otherwise it fails to build. Need to check why.

ceciliapeng2011 replied:
fixed

maxnick commented on src/plugins/intel_cpu/src/cpu_tensor.cpp:
```diff
@@ -0,0 +1,93 @@
+
```
Please remove the blank line

ceciliapeng2011 replied:
done

maxnick commented on src/plugins/intel_cpu/src/cpu_tensor.cpp:
```diff
@@ -0,0 +1,93 @@
+                std::bind1st(std::multiplies<size_t>(), m_element_type.size()));
+}
+
+void* Tensor::data(const element::Type& type) const { // GetData() call to the memory object
```
Shouldn't we perform type check? What is the purpose of the `type` parameter?

ceciliapeng2011 replied:
done

maxnick commented on src/plugins/intel_cpu/src/cpu_tensor.cpp:
```diff
@@ -0,0 +1,93 @@
+    auto blocked_desc = m_memptr->GetDescWithType<BlockedMemoryDesc>();
+    OPENVINO_ASSERT(blocked_desc);  // only support BlockedMemoryDesc so far
+
+    m_strides = blocked_desc->getStrides();
```
What is the strides format. Specifically, shall they be the same rank as the tensor? I'm asking because in the case of blocked layouts, the strides count is more then the rank.

ceciliapeng2011 replied:
Yes, ov::Tensor has Strides same rank of Shape. That means it asks plain format.

ceciliapeng2011 replied:
done

ceciliapeng2011 commented on src/plugins/intel_cpu/tests/unit/cpu_tensor_test.cpp:
```diff
@@ -0,0 +1,98 @@
+
+TEST_F(CPUTensorTest, emptySize) {
+    ov::intel_cpu::Tensor t(create_memory(Precision::FP32, {0}));
+    ASSERT_NE(nullptr, t.data());
```
This check would fail, as MemoryMngrWithReuse would never return nullptr even if required size is zero. We might force mem manager return nullptr if input size is zero. How do you think? @maxnick

maxnick replied:
AFAIK, GetData() should return nulltpr if the initial memory size is zero and the memory has not been yet resized.

ceciliapeng2011 replied:
my mistake. It is checking "ASSERT_NE" here. That means ov core expects tensor.data() returns a non-null ptr even if it is an empty tensor.
Sounds we need resize the tensor. But it still fails because of if (size > _memUpperBound) check in MemoryMngrWithReuse::resize. So I did a slight change in resize. Would you please comment?

ceciliapeng2011 replied:
done

maxnick commented on src/plugins/intel_cpu/src/cpu_tensor.h:
```diff
@@ -0,0 +1,51 @@
+
+class Tensor : public ITensor {
+public:
+    // Only plain data format is supported.
```
```suggestion
    // Only ncsp data format is supported.
```

maxnick commented on src/plugins/intel_cpu/src/cpu_tensor.cpp:
```diff
@@ -0,0 +1,101 @@
+Tensor::Tensor(MemoryPtr memptr) : m_memptr{memptr} {
+    OPENVINO_ASSERT(m_memptr != nullptr);
+
+    auto& desc = m_memptr->getDesc();
```
Suggest inserting the `ncsp` check here and remove `get_strides()` call, thus we update the strides only on demand, (i.e. when the user calls get_strides())

ceciliapeng2011 replied:
done.

maxnick commented on src/plugins/intel_cpu/src/cpu_tensor.cpp:
```diff
@@ -0,0 +1,101 @@
+
+    m_shape = std::move(new_shape);
+    m_strides.clear();
+    update_strides();
```
Better remove `update_strides()` call, again not to perform potentially unnecessary action. BTW, don't forget ot reset `m_strides_once` flag

ceciliapeng2011 replied:
std::once_flag cannot be reset.

maxnick replied:
Yep. I forgot about its semantics.

maxnick commented on src/plugins/intel_cpu/src/cpu_tensor.cpp:
```diff
@@ -0,0 +1,101 @@
+void Tensor::update_strides() const {
+    // only support plain data format so far
+    auto blocked_desc = m_memptr->GetDescWithType<BlockedMemoryDesc>();
+    OPENVINO_ASSERT(blocked_desc);
+    OPENVINO_ASSERT(blocked_desc->hasLayoutType(LayoutType::ncsp));
```
It does make sense to move the both checks to constructor, also please add error messages.

ceciliapeng2011 replied:
done

maxnick commented on src/plugins/intel_cpu/tests/unit/cpu_tensor_test.cpp:
These tests are mostly behavior tests, and this is great to have. However, we also need unit tests for the interface. To this end a mock IMemory implementation should be used and the ITensor interface is tested.

ceciliapeng2011 replied:
Since only ncsp memory is supported, what do you expect to cover with the mock IMemory impl? 
I am afraid the constructor would fail with this mock IMemory inherits, let alone inferfaces.

btw, there is no ITensor unit tests from core team, only ov::Tensor, which I referred to in this test file.

maxnick replied:
The IMemory mock class can actually have mock MemoryDesc with ncsp layout.

ceciliapeng2011 replied:
@maxnick I have initiated the unit test with Mock IMemory. Would you please check if it meets your requirements? I will update with more test cases if it is.

maxnick replied:
Yep. This is how we should test wrapper implementations.

ceciliapeng2011 committed:
impl intel_cpu::Tensor

fix

Update src/plugins/intel_cpu/src/cpu_tensor.cpp

Co-authored-by: Maksim Kutakov <maxim.kutakov@gmail.com>

Update src/plugins/intel_cpu/src/cpu_tensor.cpp

Co-authored-by: Maksim Kutakov <maxim.kutakov@gmail.com>

remove unit test of check empty tensor data().

add mock IMemory for CPUTensor unit test.

fix function template inherit issue

unit test with mock memory

update_strides fix

fix

fix

CR fix and add unit test.

trivial fix

CR fix. code cleanup.

CR fix

fix afer rebase

maxnick commented on src/plugins/intel_cpu/src/cpu_memory.cpp:
```diff
@@ -212,7 +212,7 @@ void MemoryMngrWithReuse::setExtBuff(void *ptr, size_t size) {
     constexpr int cacheLineSize = 64;
     bool sizeChanged = false;
-    if (size > m_memUpperBound) {
+    if (size > m_memUpperBound || m_data.get() == nullptr) { // conform to ov::Tensor which always has an addr even for empty tensor
```
Are we sure that ov::Tensor must return non nullptr even for an empty tensor? Allocating memory for zero size is senseless. Anyway, even if the ov::Tensor must return non-nullptr for an empty tensor, it should be processed on the ov::ITensor implementation level. Memory manager has nothing to do with the ITensor abstractions, therefore all teh ov::ITensor stuff must be located in the corresponding implementation.

ceciliapeng2011 replied:
Unluckily both them would return a non-null ptr when the size is 0.

We'd better let ITensor::data() return the same pointer of MemoryManager::resize(). Otherwise we need extra check to make sure this tensor and memory sharing the same address.

ceciliapeng2011 replied:
updated, we won't care if it return nullptr or not in case of size 0.

maxnick commented on src/plugins/intel_cpu/src/cpu_tensor.cpp:
```diff
@@ -0,0 +1,106 @@
+    m_element_type = InferenceEngine::details::convertPrecision(blocked_desc->getPrecision());
+    m_shape = blocked_desc->getShape().toPartialShape().to_shape();
+
+    if (!shape_size(m_shape)) m_memptr->getMemoryMngr()->resize(0);
```
In the case of `MemoryMngrWithReuse` this call does nothing. Moreover, we must not call resize here as in general case it may invalidate the data (in the case of other than `MemoryMngrWithReuse` memory manager implementation usage).

ceciliapeng2011 replied:
MemoryMngrWithReuse with resize(0) will do something actually. It will return a non-null ptr.
So is ov::Tensor data() with shape{0}.
Need to align with core team on the expected behavior.

ceciliapeng2011 replied:
As there is still undecisive of the expected behavior, I will delete the corresponding unit test and the code change for it.

maxnick commented on src/plugins/intel_cpu/src/cpu_tensor.cpp:
```diff
@@ -0,0 +1,106 @@
+    return desc.getCurrentMemSize();
+}
+
+const ov::Strides& Tensor::get_strides() const { // here I would ask you to investiage what solution may be better
```
Beware! This method is marked const. That means it must be thread safe!

ceciliapeng2011 replied:
This is undecisive either, as per the reply of Ilya to my email.

maxnick replied:
Every const method must be thread safe as it is assumed that it doesn't change anything. If it does, then thread safety must be insured.

ceciliapeng2011 replied:
this get_strides method would definitely change m_strides if we don't update it in set_shape and let get_strides do it, unless we don't have this m_strides member variable. 
and as Ilya's comments, it is user's responsibility to call set_shape and get_strides synchronously. that means ov::Tensor doesn't guarantee thread-safe.

Then what can we do to this ITensor implementation? 
We could add a mutex lock to set_shape and get_strides. But it would not be satisfying as we hope set_shape lighte-weighted as it might be invoked per iteration.

@maxnick What's your idea on the trade-offs?

maxnick replied:
> and as Ilya's comments, it is user's responsibility to call set_shape and get_strides synchronously. that means ov::Tensor doesn't guarantee thread-safe.

Yes, indeed. But I'm talking about synchronization between two`get_strides` calls, since this method is const. At the moment it is not safe to call `get_strides` to the same object simultaneously from two different threads. However, the user assumes such behavior to be safe since he calls a const method.
Thus we have two solutions here:
1. Add synchronization to the `get_strides` call.
2. Reset `m_strides` in the constructor and every set_shape call.
Every solution has its own tradeoffs. As a first step it looks easier to update the strides in the corresponding modifying methods.

ceciliapeng2011 replied:
done.

maxnick commented on src/plugins/intel_cpu/src/cpu_tensor.cpp:
```diff
@@ -0,0 +1,106 @@
+    // only support plain data format ncsp.
+    auto blocked_desc = m_memptr->getDescWithType<BlockedMemoryDesc>();
+    OPENVINO_ASSERT(blocked_desc, "intel_cpu::Tensor only support blocked memory.");
+    OPENVINO_ASSERT(blocked_desc->hasLayoutType(LayoutType::ncsp), "intel_cpu::Tensor only support memory with ncsp layout.");
```
```suggestion
    OPENVINO_ASSERT(blocked_desc->hasLayoutType(LayoutType::ncsp), "intel_cpu::Tensor only supports memory with ncsp layout.");
```

maxnick commented on src/plugins/intel_cpu/src/cpu_tensor.cpp:
```diff
@@ -0,0 +1,106 @@
+
+    // only support plain data format ncsp.
+    auto blocked_desc = m_memptr->getDescWithType<BlockedMemoryDesc>();
+    OPENVINO_ASSERT(blocked_desc, "intel_cpu::Tensor only support blocked memory.");
```
```suggestion
    OPENVINO_ASSERT(blocked_desc, "intel_cpu::Tensor only supports blocked memory descriptor.");
```

maxnick commented on src/plugins/intel_cpu/tests/unit/cpu_tensor_test.cpp:
```diff
@@ -91,3 +94,101 @@ TEST_F(CPUTensorTest, canSetShape) {
+    MemoryDescPtr m_pMemDesc;
+};
+
+TEST_F(CPUTensorTest, canCreateTensorUsingMockIMemory) {
```
Shouldn't it replace the original `canCreateTensor` test?

ceciliapeng2011 replied:
done

maxnick commented on src/plugins/intel_cpu/src/cpu_tensor.cpp:
```diff
@@ -0,0 +1,101 @@
+    OPENVINO_ASSERT(m_memptr != nullptr);
+
+    // only support plain data format ncsp.
+    auto blocked_desc = m_memptr->getDescWithType<BlockedMemoryDesc>();
```
This downcast is not necessary, `hasLayoutType` and `getPrecision` methods are available on the base level. Also `ncsp` layout implies that the tensor has blocked type.

ceciliapeng2011 replied:
done

maxnick commented on src/plugins/intel_cpu/src/cpu_tensor.cpp:
```diff
@@ -0,0 +1,101 @@
+// if ov::Shape is undefined (dynamic), we probably have to reintialize the internal Memory object
+// using plain memory format
+    auto& desc = m_memptr->getDesc();
+    const auto newdesc = desc.cloneWithNewDims(new_shape, true);
```
If the initial memory desc is not dynamic, this call will throw if the shape doesn't match the current static shape. Thus we have to align the behavior with the ITensor declared behavior.
If it is expected to change shape anyway, even if previously the tensor had a static shape, thus we have to create a new memory desc from scratch.
If exception is expected, than we can leave the code as it is.

ceciliapeng2011 replied:
AllocatedTensor (from src/inference/src/dev/make_tensor.cpp) which implments ITensor indicates that the implementation defines the expected behavior of set_shape. There it simply switches to a new shape without caring about the old one.
CPUTensor follows it.
As to  "thus we have to create a new memory desc from scratch", can you explain a bit why? What problem with a cloned one instead?

btw, ov::Shape could not be dynamic shape. It is always of static shape semantically, though some dim of it could be zero.

maxnick replied:
> As to "thus we have to create a new memory desc from scratch", can you explain a bit why? What problem with a cloned one instead?

Actually I didn't recognize the `relaxedCheck` flag behavior. Initially we designed this method so that it doesn't allow to clone with new dims if they are not compatible with the existing shape. Ok then, let's leave it as it is.
> 
> btw, ov::Shape could not be dynamic shape. It is always of static shape semantically, though some dim of it could be zero.

Good observation. Let's throw in the case of attempt to get shape from a dynamic shape memory for the beginning. We can always relax this condition later.

maxnick commented on src/plugins/intel_cpu/src/cpu_tensor.cpp:
```diff
@@ -0,0 +1,101 @@
+    OPENVINO_ASSERT(blocked_desc, "intel_cpu::Tensor only supports blocked memory descriptor.");
+    OPENVINO_ASSERT(blocked_desc->hasLayoutType(LayoutType::ncsp), "intel_cpu::Tensor only supports memory with ncsp layout.");
+
+    m_element_type = InferenceEngine::details::convertPrecision(blocked_desc->getPrecision());
+    m_shape = blocked_desc->getShape().toPartialShape().to_shape();
```
What if the underlying memory object descriptor is changed after the `Tesnor` object has been instanced? Then the shape change is not reflected by the `Tensor` object. Suggest moving this lines to the corresponding methods. This wrapped doesn't solely own the memory object, and it may be modified somewhere outside the object as well. The same is true for the strides.

ceciliapeng2011 replied:
Similar to strides, the places where m_shape would change is set_shape. It resets m_shape there. So there is no problem to it.

maxnick replied:
The problem is still here. What if the underlying memory object memory descriptor was changed in a separate code frame? How the shape and the strides are updated. The `Tensor` object stores only a reference to the memory, thus the memory itself is not protected from modifications outside the `Tensor` object. Also this indicates  a lack of tests. Simple test could be as follows:

```cpp
    Memory mem(eng, desc);
    Tensor tensor(mem);
    ASSERT_EQ(mem.getDesc().getStrides() * data_size, tensor.get_strides()); // here we assume desc.getStrides() * data_size is a scaling of the whole vector
    auto desc2 = desc.cloneWithNewDims(newShape);
    mem.redefineDesc(desc2);
    ASSERT_EQ(mem.getDesc().getStrides() * data_size, tensor.get_strides()); // this check fails since the current implementation doesn't reflect the change of the underlining memory object state
```

ceciliapeng2011 replied:
fixed and updated with more unit tests.

maxnick commented on src/plugins/intel_cpu/tests/unit/cpu_tensor_test.cpp:
```diff
@@ -0,0 +1,229 @@
+
+    MOCK_METHOD(bool, hasLayoutType, (LayoutType), (const, override));
+
+    // MOCK_METHOD(std::string, serializeFormat, (), (const, override));
```
Commented code.

ceciliapeng2011 replied:
done

ceciliapeng2011 commented on src/plugins/intel_cpu/src/cpu_tensor.cpp:
```diff
@@ -0,0 +1,101 @@
+    OPENVINO_ASSERT(blocked_desc->hasLayoutType(LayoutType::ncsp), "intel_cpu::Tensor only supports memory with ncsp layout.");
+
+    m_element_type = InferenceEngine::details::convertPrecision(blocked_desc->getPrecision());
+    m_shape = blocked_desc->getShape().toPartialShape().to_shape();
```
> m_shape = blocked_desc->getShape().toPartialShape().to_shape();

discussion:
ov::Tensor is always of static shape (which is seen from ov::Shape semantics), though it could be of zero dims.
but the argument memptr could be constructed with a partialshape and it could be of dynamic shape.

Then this call to to_shape() would fail if it is a memory with dynamic shape.

Question: what should we do? 
- 1. Not allow memory ptr with dynamic shape in this constructor?  
- 2. Or, allow it but throw exception in get_shape/get_strides if it is not static shape?

I have no bias to the options. What's your opinion? @maxnick

maxnick replied:
Please see https://github.com/maxnick/openvino/pull/63#discussion_r1243290479
Suggest going with the second option

ceciliapeng2011 replied:
Sounds your option is 1, based on your comment https://github.com/maxnick/openvino/pull/63#discussion_r1243290479

ceciliapeng2011 replied:
updated with dynamic shape checks in constructor, get_shape, get_strides, etc.. as memptr might be changed to dynamic shape anytime out of boundary of ITensor.

maxnick replied:
> Good observation. Let's throw in the case of attempt to get shape from a dynamic shape memory for the beginning. We can always relax this condition later.

There is nothing about the constructor. We shouldn't thow in the constructor, since we can initialize the Tensor with a memory of dynamic shape, and then call `set_shape` with a defined shape before calling `get_shape` or `get_strides`. So only those methods should check dynamic dimensions.

ceciliapeng2011 replied:
done

maxnick commented on src/plugins/intel_cpu/src/cpu_tensor.cpp:
```diff
@@ -0,0 +1,104 @@
+    // only support plain data format ncsp.
+    auto memdesc = m_memptr->getDescPtr();
+    OPENVINO_ASSERT(memdesc->hasLayoutType(LayoutType::ncsp), "intel_cpu::Tensor only supports memory with ncsp layout.");
+    OPENVINO_ASSERT(m_memptr->getDescPtr()->isDefined(), "intel_cpu::Tensor requires memory with defined descriptor.");
```
We should allow the Tensor object to be constructed from dynamic memory, see https://github.com/maxnick/openvino/pull/63#discussion_r1244927191

ceciliapeng2011 replied:
done

maxnick commented on src/plugins/intel_cpu/src/cpu_tensor.cpp:
```diff
@@ -0,0 +1,104 @@
+
+const ov::Shape& Tensor::get_shape() const { //trivial -> convert cpu Shape to ov::Shape
+    auto& shape = m_memptr->getDescPtr()->getShape();
+    OPENVINO_ASSERT(shape.isStatic(), "intel_cpu::Tensor requires memory initialized with static shape.");
```
```suggestion
    OPENVINO_ASSERT(shape.isStatic(), "intel_cpu::Tensor has dynamic shape.");
```

ceciliapeng2011 replied:
done

maxnick commented on src/plugins/intel_cpu/src/cpu_tensor.cpp:
```diff
@@ -0,0 +1,104 @@
+    auto& shape = m_memptr->getDescPtr()->getShape();
+    OPENVINO_ASSERT(shape.isStatic(), "intel_cpu::Tensor requires memory initialized with static shape.");
+
+    std::lock_guard<std::mutex> guard(m_lock);
+    m_shape = shape.toPartialShape().to_shape();
+    return m_shape;
```
What is the sense to save the shape inside the object, `m_shape` is not used anyhow? Why cant we simply `return shape.toPartialShape().to_shape();`? Also it allow to avoid using the lock. Also probably we can skip the toPartialShape() stage, and initialize the ov::Shape from static shape, something like that `return {shape.getStaticDims()}`

ceciliapeng2011 replied:
It is illegal to return reference to a local variable. That's why m_shape is used. Sounds no good way to get a ov::Shape reference from intel_cpu::Shape.

maxnick replied:
Yes, does make sense, didn't notice that the return type is a reference. However using `toPartialShape` is still an unnecessary step. It is better to use `m_shape = ov::Shape{shape.getStaticDims()}` as I've mentioned above.

ceciliapeng2011 replied:
done

maxnick commented on src/plugins/intel_cpu/src/cpu_tensor.cpp:
```diff
@@ -0,0 +1,104 @@
+// perhaps we are not allowed to wrap any cpu Memory except built over BlockedMemoryDesc
+    OPENVINO_ASSERT(m_memptr->getDescPtr()->isDefined(), "intel_cpu::Tensor requires memory with defined strides.");
+
+    std::lock_guard<std::mutex> guard(m_lock);
+    update_strides();
+    return m_strides;
```
The same here, we do not need to store `m_strides` by any reason. We can simply read them from the descriptor, modify and return in the required format.

ceciliapeng2011 replied:
better keep m_strides as get_strides returns "const ov::Strides&". The same to m_shape.

maxnick replied:
Ok got it.

maxnick commented on src/plugins/intel_cpu/src/cpu_tensor.cpp:
```diff
@@ -0,0 +1,103 @@
+}
+
+void Tensor::update_strides() const {
+    auto blocked_desc = m_memptr->getDescWithType<BlockedMemoryDesc>();
```
It is still beter to have null check, just in case the mem descriptor type  in the memory object has been completely changed by some reason.

ceciliapeng2011 replied:
done

maxnick commented on src/plugins/intel_cpu/src/cpu_tensor.h:
```diff
@@ -0,0 +1,46 @@
+
+    void* data(const element::Type& type = {}) const override;
+
+protected:
```
protected -> private

ceciliapeng2011 replied:
done

maxnick reviewed (commented):
Almost done. Please clean up unnecessary comments in the cpu_tensor.cpp as well as the commented code in the test part.

maxnick commented:
Close as the code has been already merged to the main repo.
